keps/sig-node/4603-tune-crashloopbackoff/README.md

```diff
@@ -0,0 +1,1431 @@
+unreasonable or unpredictable defaults. Therefore the approach manages the risk
+of different decay curves to node stability by providing an API surface to
+opt-in to the riskiest option, and ultimately also provides the observability
+instrumentation needed by cluster operators to model the impact of this change
+on their existing deployments.
+
+A large number of alternatives have been discussed over the 5+ years the
+canonical tracking issue has been open, some of which imply high levels of
+sophistication for kubelet to make different decisions based on system state,
+workload specific factors, or by detecting anomalous workload behavior. While
+this KEP does not rule out those directions in the future, the proposal herein
+focuses on the simpler, easily modeled changes that are designed to address the
+most common issues observed today. It is the desire of the author that the
+observability and benchmarking improvements instrumented as part of this
+proposal can serve as a basis for pursuing those types of improvements in the
+future, including as signals for end users to represent such solutions as
+necessary. Some analysis of these alternatives are included in the Alternatives
+Considered section below for future reference.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+* Improve pod restart backoff logic to better match the actual load it creates
+  and meet emerging use cases
+* Quantify and expose risks to node stability resulting from decreased and/or
+  hetereogeneous backoff configurations
+* Empirically derive defaults and allowable thresholds that approximate current
+  node stability
+* Provide a simple UX that does not require changes for the majority of
+  workloads
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+* This effort is NOT intending to support fully user-specified configuration, to
+  cap risk to node stability
+* This effort is purposefully NOT implementing more complex heuristics by
+  kubelet (e.g. system state, workload specific factors, or by detecting
+anomalous workload behavior) to focus on better benchmarking and observability
+  and address common use cases with easily modelled changes first
+* This effort is NOT changing the independent backoff curve for image pulls
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+This design seeks to incorporate a two-pronged approach:
+
+1. Change the existing initial value for the backoff curve to stack more retries
```

Comment:
I wonder if both of these options shouldn't be a cluster-admin level configs. This way the operator is able to ensure the acceptable load on the cluster. Otherwise we might end up in the situation where a cluster gets DDoS-ed with too many Rapid pods, which I believe you're planning to exercise during alpha perf testing.

Reply:
> I wonder if both of these options shouldn't be a cluster-admin level configs

Do you mean something different from the alpha feature gates (or perhaps, the feature gates being configurable forever?)

For Rapid: I do want each individual `Rapid` to be configured by Pod as every use case has had a specific Pod or sidecar in mind. Or did you mean something else here?

It is true I am planning to exercise the DDoS problem during alpha perf testing. Really what I've selected are starting values that we could generally agree on from initial testing and anecdotal experience for alpha use, but I expect the actual values to change during the systematic performance testing in the alpha perf testing period.

For the baseline curve: Do you mean that, in perpetuity, cluster admins can choose their own cluster-wise initial values (or, at least, between today's initial of 10s and the alpha initial of 1s)?